node/actors/network/src/gossip/attestation/mod.rs

```diff
@@ -0,0 +1,330 @@
+        };
+        if self.votes.contains_key(&vote.key) {
+            return Ok(());
+        }
+        // Verify signature only after checking all the other preconditions.
+        vote.verify().context("verify")?;
+        self.votes.insert(vote.key.clone(), vote);
+        self.weight += weight;
+        Ok(())
+    }
+
+    fn insert_votes(
+        &mut self,
+        votes: impl Iterator<Item = Arc<attester::Signed<attester::Batch>>>,
+    ) -> anyhow::Result<()> {
+        let mut done = HashSet::new();
+        for vote in votes {
+            // Disallow multiple entries for the same key:
+            // it is important because a malicious attester may spam us with
+            // new versions and verifying signatures is expensive.
+            if done.contains(&vote.key) {
+                anyhow::bail!("duplicate entry for {:?}", vote.key);
+            }
+            done.insert(vote.key.clone());
+            self.insert_vote(vote)?;
+        }
+        Ok(())
+    }
+
+    fn qc(&self) -> Option<attester::BatchQC> {
+        if self.weight < self.config.committee.threshold() {
+            return None;
+        }
+        let mut sigs = attester::MultiSig::default();
+        for vote in self.votes.values() {
+            sigs.add(vote.key.clone(), vote.sig.clone());
+        }
+        Some(attester::BatchQC {
+            message: self.config.batch_to_attest.clone(),
+            signatures: sigs,
+        })
+    }
+}
+
+/// Receiver of state diffs.
+pub(crate) struct DiffReceiver {
+    prev: Option<State>,
+    recv: sync::watch::Receiver<Option<State>>,
+}
+
+impl DiffReceiver {
+    /// Waits for the next state diff.
+    pub(crate) async fn wait_for_diff(&mut self, ctx: &ctx::Ctx) -> ctx::OrCanceled<Diff> {
+        loop {
+            let Some(new) = (*sync::changed(ctx, &mut self.recv).await?).clone() else {
+                continue;
+            };
+            let diff = new.diff(&self.prev);
+            self.prev = Some(new);
+            if !diff.is_empty() {
+                return Ok(diff);
+            }
+        }
+    }
+}
+
+/// `Controller` manages the attestation state.
+/// It maintains a set of votes matching the attestation config.
+/// It allows for
+/// * adding votes to the state
+/// * subscribing to the vote set changes
+/// * waiting for the certificate to be collected
+///
+/// It also keeps an attester key used to sign the batch vote,
+/// whenever it belongs the current attester committee.
+/// Signing happens automatically whenever the committee is updated.
+///
+/// Expected usage:
+/// ```
+/// let ctrl = attestation::Controller::new(Some(key));
+/// loop {
+///     // Check the global attestation registry.
+///     // Compute the next expected batch and the committee that should attest it.
+///     ...
+///     let config = attestation::Config {
+///         batch_to_attest: ...,
+///         committee: ...,
+///     };
+///     ctrl.update_config(Arc::new(config.clone())).unwrap();
```

Comment:
IMO this is a very confusing interface. Why is the method that communicates the next batch details to the consensus subsystem called `update_config`? Also `Config` is not a good name for a struct belonging to the public interface

Something like
```
  attester_controller.start_batch_attestation(BatchAttesterInfo {number, committe})
```

Reply:
start_batch_attestation, ok. But how is "Info" better than "Config"? Info doesn't even indicate whether data is input or output of the system.

Reply:
This is why I suggested that `update` could take multiple parameters and the `Config` could be an internal detail, perhaps?

Reply:
yes, I'll be much more comfortable with `Config` being internal to consensus. 
`Info` is better because it's more generic - `Config` associates with something configurable, in most cases configurable by operator. This is not the case here. as @aakoshh mentioned elsewhere, this name is common in literature, but not everyone working on era code has this background

Reply:
I'm not convinced, but ok, `Info` it is.